client/shared/src/main/scala/org/http4s/client/Client.scala

```diff
@@ -289,16 +291,40 @@ object Client {
       go(source).stream
     }
 
-    def runOrDispose(req: Request[F]): F[Resource[F, Response[F]]] =
-      Ref[F].of(false).map { disposed =>
-        val req0 = addHostHeaderIfUriIsAbsolute(req.pipeBodyThrough(until(disposed)))
-        Resource
-          .eval(app(req0))
-          .onFinalize(disposed.set(true))
-          .map(_.pipeBodyThrough(until(disposed)))
-      }
+    def run(
+        message: Message[F],
+        refOp: Option[Ref[F, Boolean]] = None,
+    ): Resource[F, Response[F]] = message match {
+      case req: Request[F] =>
+        Resource.eval(Ref[F].of(false)).flatMap { disposed =>
+          val reqAugmented =
+            addHostHeaderIfUriIsAbsolute(req.pipeBodyThrough(until(disposed)))
+          Resource
+            .eval(app(reqAugmented))
+            .onFinalize(disposed.set(true))
+            .flatMap(run(_, Option(disposed)))
+        }
+      case resp: Response[F] =>
+        for {
+          disposed <- Resource.eval(refOp.fold(Ref[F].of(false))(F.pure))
+          channel <- Resource.make(Channel.unbounded[F, Byte])(_.close.void)
```

Comment:
I think this should be a `Channel.synchronous`. Otherwise the entire response will go into memory, even before the client has requested it, so we lose backpressure. A `synchronous` stream will only request when the downstream has requested.

Also, `channel.sendAll(...)` closes the `Channel` automatically. Instead I feel like the `close` here should be draining the `Channel`.